src/main/java/org/skriptlang/skript/bukkit/recipes/elements/ExprRecipeIngredients.java

```diff
@@ -0,0 +1,399 @@
+									yield trimRecipe.getTemplate();
+								yield null;
+							}
+							case ADDITION -> smithingRecipe.getAddition();
+                            default -> null;
+                        };
+						if (choice instanceof ExactChoice exactChoice) {
+							ingredients.addAll(exactChoice.getChoices());
+						} else if (choice instanceof MaterialChoice materialChoice) {
+							ingredients.addAll(
+								materialChoice.getChoices().stream().map(ItemStack::new).toList()
+							);
+						}
+					} else {
+						//Skript.error("You can only get the base, template, and addition items of a Smithing, Smithing Transform and Smithing Trim Recipe.");
+					}
+				}
+				case INPUT -> {
+					if (recipe instanceof MutableRecipe mutableRecipe) {
+						if (mutableRecipe instanceof MutableCookingRecipe mutableCookingRecipe) {
+							ingredients.addAll(((ExactChoice) mutableCookingRecipe.getInput()).getChoices());
+						} else if (mutableRecipe instanceof MutableStonecuttingRecipe mutableStonecuttingRecipe) {
+							ingredients.addAll(((ExactChoice) mutableStonecuttingRecipe.getInput()).getChoices());
+						}
+					} else {
+						RecipeChoice choice = null;
+						if (recipe instanceof CookingRecipe<?> cookingRecipe) {
+							choice = cookingRecipe.getInputChoice();
+						} else if (recipe instanceof StonecuttingRecipe stonecuttingRecipe) {
+							choice = stonecuttingRecipe.getInputChoice();
+						} else {
+							//Skript.error("You can only get the input item of a Cooking, Blasting, Furnace, Campfire, Smoking and Stonecutting Recipe.");
+						}
+						if (choice instanceof ExactChoice exactChoice) {
+							ingredients.addAll(exactChoice.getChoices());
+						} else if (choice instanceof MaterialChoice materialChoice) {
+							ingredients.addAll(materialChoice.getChoices().stream().map(ItemStack::new).toList());
+						}
+					}
+				}
+			}
+		}
+		return ingredients.toArray(ItemStack[]::new);
+	}
+
+	@Override
+	public Class<ItemStack> getReturnType() {
+		return ItemStack.class;
+	}
+
+	@Override
+	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
+		if (!isEvent) {
+			Skript.error("You can not set the " + selectedChoice.toString + " of existing recipes.");
+		} else {
+			if (mode == ChangeMode.SET) {
+				return CollectionUtils.array(ItemType[].class);
+			}
+		}
+		return null;
+	}
+
+	@Override
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
+		if (!(event instanceof RegisterRecipeEvent recipeEvent))
+			return;
+
+		Map<Integer, ItemStack[]> items = new HashMap<>();
+		for (int i = 0; i < delta.length; i++) {
+			Object object = delta[i];
+			if (object instanceof ItemType itemType) {
+				List<ItemStack> thisItems = new ArrayList<>();
+				itemType.getAll().forEach(thisItems::add);
+				items.put(i, thisItems.toArray(ItemStack[]::new));
+			}
+		}
+
+		MutableRecipe mutableRecipe = recipeEvent.getMutableRecipe();
+
+		switch (selectedChoice) {
```

Comment:
again
needs to be split up
if your switch statement is over 20 lines, it should probably be helper methods